validation/gwas/method/regenie/glow_wgr.py

```diff
@@ -0,0 +1,381 @@
+        .load(plink_path)
+    )
+
+    variant_df = df.withColumn(
+        "values", mean_substitute(genotype_states(F.col("genotypes")))
+    ).filter(F.size(F.array_distinct("values")) > 1)
+    if contigs is not None:
+        variant_df = variant_df.filter(F.col("contigName").isin(contigs))
+
+    sample_ids = get_sample_ids(variant_df)
+    logger.info(f"Found {len(sample_ids)} samples, first 10: {sample_ids[:10]}")
+
+    ###########
+    # Stage 1 #
+    ###########
+
+    logger.info(HR)
+    logger.info("Calculating variant/sample block info")
+    block_df, sample_blocks = block_variants_and_samples(
+        variant_df,
+        sample_ids,
+        variants_per_block=variants_per_block,
+        sample_block_count=sample_block_count,
+    )
+
+    label_df = pd.read_csv(traits_path, index_col="sample_id")
+    label_df = (label_df - label_df.mean()) / label_df.std(ddof=0)
+    logger.info(HR)
+    logger.info("Trait info:")
+    logger.info(_info(label_df))
+
+    cov_df = pd.read_csv(covariates_path, index_col="sample_id")
+    cov_df = (cov_df - cov_df.mean()) / cov_df.std(ddof=0)
+    logger.info(HR)
+    logger.info("Covariate info:")
+    logger.info(_info(cov_df))
+
+    stack = RidgeReducer(alphas=alphas)
+    reduced_block_df = stack.fit_transform(block_df, label_df, sample_blocks, cov_df)
+    logger.info(HR)
+    logger.info("Stage 1: Reduced block schema:")
+    logger.info(_schema(reduced_block_df))
+
+    path = output_path / "reduced_blocks.parquet"
+    reduced_block_df.write.parquet(str(path), mode="overwrite")
+    logger.info(f"Stage 1: Reduced blocks written to {path}")
+
+    # Flatten to scalars for more convenient access w/o Spark
+    flat_reduced_block_df = spark.read.parquet(str(path))
+    path = output_path / "reduced_blocks_flat.csv.gz"
+    flat_reduced_block_df = _flatten_reduced_blocks(flat_reduced_block_df)
+    flat_reduced_block_df = flat_reduced_block_df.toPandas()
+    flat_reduced_block_df.to_csv(path, index=False)
+    # flat_reduced_block_df.write.parquet(str(path), mode='overwrite')
+    logger.info(f"Stage 1: Flattened reduced blocks written to {path}")
+
+    ###########
+    # Stage 2 #
+    ###########
+
+    # Monkey-patch this in until there's a glow release beyond 0.5.0
+    if glow_version != "0.5.0":
+        raise NotImplementedError(
+            f"Must remove adjustements for glow != 0.5.0 (found {glow_version})"
+        )
+    # Remove after glow update
+    RidgeRegression.transform_loco = transform_loco
+    estimator = RidgeRegression(alphas=alphas)
+    model_df, cv_df = estimator.fit(reduced_block_df, label_df, sample_blocks, cov_df)
+    logger.info(HR)
+    logger.info("Stage 2: Model schema:")
+    logger.info(_schema(model_df))
+    logger.info("Stage 2: CV schema:")
+    logger.info(_schema(cv_df))
+
+    y_hat_df = estimator.transform(
+        reduced_block_df, label_df, sample_blocks, model_df, cv_df, cov_df
+    )
+
+    logger.info(HR)
```

Comment:
I guess this is a validation code, so maybe this isn't too important, but somethings you use `logger` and sometimes `print`, it might be better to just standardise on one (log?)

Reply:
Yea, I suppose the pyinvoke tasks could use a logger too.  I was hoping this kind of nitpick wouldn't be necessary for the test data generation scripts but I'll update it.

Reply:
https://github.com/pystatgen/sgkit/pull/66/commits/64c1da651f5cd2596f28461639981809f80519ad#diff-7240a770aa7f50f89b47ef6286c19a81R11